[OP#40868] development environment with docker compose (#42)
nextcloud/integration_openproject: README.md

#### Background jobs

To be able to periodically check activity in OpenProject (when "notifications for activity in my work packages" is enabled), you need to choose the "Cron" background job method and set a system cron task calling cron.php as explained in the [documentation](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/background_jobs_configuration.html#cron).

## Development
Develop using docker compose

Requirements:
- Node.js
- Docker, Docker Compose
- OpenProject server instance running in the host machine
- OpenProject Integration app

### Environments
- **APP_DIR**
  - description: location where the `integration_openproject` repository is cloned 
  - default: `./../integration_openproject`

### Start compose

> **Note:** If your host has running apache or nginx, you need to stop it before starting the `docker-compose`.

Reviewer: can't we use an alternative port inside of docker, so it would not affect any server on the host?

Author: when using the network mode as `host`, we cannot do port mapping

```shell
APP_DIR='/path/to/app' docker-compose up -d
```

After this, you should be able to access nextcloud server at [http://localhost](http://localhost).

### Setup NC server

> **Note:** These steps will only be necessary for the first setup.

#### Fix permissions
User `www-data` should have _write_ permission to modify `apps` & `custom_apps` directories inside the server directory.

```shell
# "integration_openproject_nc" is the name of our nc server container
docker exec integration_openproject_nc chown www-data custom_apps/ -R
```

#### Create admin
For the database, **PostgreSQL** is used with the following credentials:
- **Database:** `nextcloud`
- **User:** `nextcloud`
- **Password:** `nextcloud`

1. browse to [http://localhost](http://localhost)
2. create admin user
3. get an installed NC server

#### Enable integration app 
You can browse as admin to the apps center and enable it using the webUI. Or, you can just use the terminal as:

```shell
docker exec --user www-data integration_openproject_nc php occ a:e integration_openproject
```

If you want to allow local remote servers: 

```shell

> docker exec --user www-data integration_openproject_nc php occ config:system:set allow_local_remote_servers --value 1

Reviewer: can that be done automatically?

Author: NC server should be in installed state, so we've to do it manually

- OpenProject server instance running in the host machine
- OpenProject Integration app

### Setup
```shell
# the app needs to be cloned inside the "custom_apps" dir anywhere in the host

> mkdir /dev/custom_apps

Reviewer: this is confusing because there is an actual `/dev` (for devices) folder on linux
```suggestion
mkdir ~/development/custom_apps
```

> git clone https://github.com/nextcloud/integration_openproject.git

Reviewer: ```suggestion
cd ~/development/custom_apps
git clone https://github.com/nextcloud/integration_openproject.git
```

# installation & building
npm ci
npm run build

> # provide ownership of "custom_apps" to the user "www-data"
> sudo chown www-data custom_apps -R

Reviewer: if you do that, then the developer will not have write access to the app anymore
I would suggest to give it to the group `www-data` and give write access to the group
```suggestion
# provide group ownership of "custom_apps" to the user "www-data"
sudo chgrp www-data ~/development/custom_apps -R
sudo chmod g+w ~/development/custom_apps -r
```

docker exec --user www-data integration_openproject_nc php occ a:e integration_openproject
```

#### Allow local remote servers: 

```shell
docker exec --user www-data integration_openproject_nc php occ config:system:set allow_local_remote_servers --value 1
```

### Start Developing
Now you can watch for the app code changes using the following command and start developing.

```shell

> cd /dev/custom_apps/integration_openproject

Reviewer: ```suggestion
cd ~/development/custom_apps/integration_openproject
```
